Picasso

manuelmeister opened the pull request:
Fixes #494 
Fixes #502

manuelmeister committed:
Added drag and drop

manuelmeister committed:
Fixed key from add to new

manuelmeister committed:
Added local scheduleEntries and added popover v-menu

manuelmeister committed:
Extracted new entry creation

manuelmeister committed:
Extracted entry change

manuelmeister committed:
Extracted entry move

manuelmeister committed:
Renamed moveEntry to moveEntryTime

manuelmeister committed:
Faster loading

manuelmeister committed:
Added watcher

manuelmeister committed:
Extracted clear entries

manuelmeister committed:
Decide when to show EntryInfoPopup

manuelmeister committed:
Show EntryPopup

manuelmeister committed:
Added cancel entry popup

manuelmeister committed:
Added activityCategory()
Revert entry back to original on cancel

carlobeltrame commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -24,16 +24,16 @@ Show all activity schedule entries of a single period.
         <picasso
-          v-else
-          :camp="camp"
+          v-show="!listFormat"
```
https://www.digitalocean.com/community/tutorials/vuejs-conditional-directives#v-show
> Usually, v-show has a performance advantage if the elements are switched on and off frequently, while the v-if has the advantage when it comes to initial render time.

Warum nicht mehr `v-if`? Und falls du doch wieder `v-if` verwendest, könntest du die Änderungen in diesem File komplett reverten.

manuelmeister committed:
Working drag and drop with popup

manuelmeister committed:
Don't move on middle mouseclick activity open

manuelmeister committed:
Removed jank and added possibility to open activity via ctrl and cmd key

manuelmeister committed:
Switched to $tc

manuelmeister committed:
Merge branch 'devel' into feature/picasso

manuelmeister committed:
Open activity from popup

manuelmeister committed:
Organized code

manuelmeister committed:
Updated snapshots

manuelmeister committed:
Allow longer duration, added min-width for the days and optimized date format in day header

manuelmeister committed:
Added now line

manuelmeister committed:
Improved coloring for past events

manuelmeister committed:
Merge branch 'devel' of github.com:ecamp/ecamp3 into feature/picasso

manuelmeister committed:
First try implementing period into camp url

manuelmeister commented:
@carlobeltrame #502 habe versucht die Periode in die URL zu übernehmen, aber das prop **period** bleibt `undefined`. Kannst du mal den letzten Commit 903d524 anschauen?

carlobeltrame committed:
Pass on all attributes as props from camp component to detail component

carlobeltrame committed:
Fix linking to camp program

Don't nest admin and collaborators under the period (for now).
Instead, dynamically and asynchronously redirect from the camp/program
route to the camp/period route, using the first non-finished period.
Refs #502

carlobeltrame committed:
Make periods in sidebar clickable

carlobeltrame commented:
Das Problem war dass in der Camp.vue Komponente die `period` nicht an die CampProgram.vue Komponente weitergegeben wurde. Ich habe es jetzt so geändert dass automatisch alle Props weitergegeben werden.
Habe auch gleich den Rest von #502 implementiert und auf deinen Branch gepusht.

manuelmeister committed:
Print Background Job

manuelmeister committed:
Implemented create activity with a schedule entry

manuelmeister committed:
Remove facebook

manuelmeister committed:
Added patch

manuelmeister committed:
Removed prop schedule entries

manuelmeister committed:
Remove scheduleEntries prop in CampProgram.vue

manuelmeister committed:
Removed prop camp from picasso

manuelmeister committed:
Removed camp from props

manuelmeister committed:
Make db accessible from outside, so you can use db-viewers

manuelmeister committed:
Use milliseconds timestamp as default time format

manuelmeister committed:
Fixed wrong assert value

manuelmeister committed:
Allow to patch ScheduleEntry with utc timestamp

pmattmann commented on backend/module/eCampCore/src/EntityService/ScheduleEntryService.php:
```diff
@@ -38,4 +43,32 @@ protected function fetchQueryBuilder($id) {
+
+        /** @var ScheduleEntry $scheduleEntry */
+        $scheduleEntry = parent::createEntity($data);
+        $scheduleEntry->setPeriod($period);
```
Use $period->addScheduleEntry($scheduleEntry);

pmattmann commented on backend/module/eCampCore/src/EntityService/ScheduleEntryService.php:
```diff
@@ -38,4 +43,32 @@ protected function fetchQueryBuilder($id) {
+        /** @var ScheduleEntry $scheduleEntry */
+        $scheduleEntry = parent::createEntity($data);
+        $scheduleEntry->setPeriod($period);
+        $scheduleEntry->setActivity($activity);
```
Use $activity->addScheduleEntry($scheduleEntry)

pmattmann reviewed (dismissed):
#519

manuelmeister committed:
Changed from set parent to add child as defined in issue #519

manuelmeister committed:
Added PHP Docs to inform that the parent is not set directly

manuelmeister committed:
Allow store objects in data in api post call. This simplifies the creation of new objects which reference existing objects.
They won't get sent over the network to minimize payload. Instead they are reduced to
```js
{ [name + 'Id']: id }
```

manuelmeister committed:
- Allow ActivityService to create ScheduleEntries and reverse.
- ScheduleEntry `start` is now called `periodOffset`

manuelmeister committed:
Missing refactor from start to periodOffset

manuelmeister committed:
Made ScheduleEntryService a lazy service

manuelmeister committed:
Create activity if activity is set

manuelmeister committed:
Changed start- & endTime to be calculated on the fly based on the period, to respect the principle that ScheduleEntries are not bound to time but to the offset in minutes of their respective period

manuelmeister committed:
Allow for multiple projects to be open

manuelmeister committed:
Merge remote-tracking branch 'upstream/devel' into feature/picasso

# Conflicts:
#	frontend/src/views/activity/Activity.vue

manuelmeister committed:
Fixed tests

manuelmeister committed:
Merged language for activity

manuelmeister committed:
Added location, as it is now a required field on activity

manuelmeister committed:
Added errorhandling on creation

manuelmeister commented:
Doesn't work fully. Main focus was on activity creation. Next tasks would be to split up the picasso because of its enormous size and existing ScheduleEntry modification

manuelmeister committed:
Fixed wrong name

manuelmeister committed:
Show dialog on click on ScheduleEntry

manuelmeister committed:
Moved ScheduleEntryPopup out of Picasso

manuelmeister committed:
Added linebreak in code

manuelmeister committed:
Add save method

manuelmeister committed:
Do not remove ScheduleEntry from list upon opening of popup

manuelmeister committed:
Only send relevant data

manuelmeister committed:
Fixed new activity not actually including the camp

manuelmeister committed:
Evaluate value if it is a function

manuelmeister committed:
Linted

manuelmeister commented:
### Meeting Notes
- use Dialog for ScheduleEntryPopup

manuelmeister committed:
Added format and parse to timepicker

manuelmeister committed:
Allow activityCategory of activity to be patched

manuelmeister committed:
Used Dialog instead of ScheduleEntryPopup.vue

manuelmeister committed:
Removed lazy service again

manuelmeister committed:
Removed ActivityService from ScheduleEntryService

manuelmeister committed:
Fixed bugs

manuelmeister committed:
Removed unnecessary loading pipes

manuelmeister committed:
Prematurely removed loading pipes

manuelmeister committed:
Added language

manuelmeister committed:
Merge branch 'devel' of github.com:ecamp/ecamp3 into feature/picasso

manuelmeister committed:
Fixed merge conflicts, missing strings and wrong translations.

manuelmeister committed:
Updated composer

manuelmeister committed:
Revert "Updated laminas-dependency-plugin because it threw an error while building with composer v2"

This reverts commit c1ce3ec5272f6f13907598ce3a8dbc393d9c1e38.

manuelmeister committed:
Updated laminas-dependency-plugin because it threw an error while building with composer v2

manuelmeister commented:
@carlobeltrame the e2e test doesn't work anymore. Are you on it, or should I try to solve it?

carlobeltrame commented:
Looks like a temporary Travis issue to me, I had a similar one a few days ago on a different project. Locally the commands work as expected, and the "failing" command is just waiting on the containers to bind to the network interface, which can't really go wrong just because of our code.

In the meantime, when I check out your branch and try to run the containers, the composer container yields the following error:
```
ecamp3-composer           |   Problem 1
ecamp3-composer           |     - Installation request for ocramius/proxy-manager 2.9.1 -> satisfiable by ocramius/proxy-manager[2.9.1].
ecamp3-composer           |     - ocramius/proxy-manager 2.9.1 requires composer-runtime-api ^2.0.0 -> no matching package found.
ecamp3-composer           |   Problem 2
ecamp3-composer           |     - doctrine/migrations 2.2.1 requires ocramius/proxy-manager ^2.0.2 -> satisfiable by ocramius/proxy-manager[2.9.1].
ecamp3-composer           |     - doctrine/migrations 2.2.1 requires ocramius/proxy-manager ^2.0.2 -> satisfiable by ocramius/proxy-manager[2.9.1].
ecamp3-composer           |     - ocramius/proxy-manager 2.9.1 requires composer-runtime-api ^2.0.0 -> no matching package found.
ecamp3-composer           |     - Installation request for doctrine/migrations 2.2.1 -> satisfiable by doctrine/migrations[2.2.1].
ecamp3-composer           | 
ecamp3-composer           | Potential causes:
ecamp3-composer           |  - A typo in the package name
ecamp3-composer           |  - The package is not available in a stable-enough version according to your minimum-stability setting
ecamp3-composer           |    see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
ecamp3-composer           |  - It's a private package and you forgot to add a custom repository to find it
ecamp3-composer           | 
ecamp3-composer           | Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.
ecamp3-composer exited with code 2
```

This indicates to me that you have updated the dependencies a bit too far, because they require composer v2, while we still have composer 1.10 defined in docker-compose.yml.

When I try to auto-fix composer.lock using `docker-compose run composer update`, I get the following errors:
```
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package laminas/laminas-dependency-plugin ^2.0 is satisfiable by laminas/laminas-dependency-plugin[2.0.0beta1, 2.0.x-dev] but these conflict with your requirements or minimum-stability.
```

laminas-dependency-plugin v2.0 is not yet stable, so composer refuses to upgrade to it. You need to downgrade, or remove it entirely (see also #555).

manuelmeister commented:
I think the problem lays with travis builds. The backend build seems to self-update composer. I don't know where this update comes, otherwise I would fix it.
Otherwise my pull request would be ready for review.

manuelmeister committed:
Merge branch 'devel' into feature/picasso

manuelmeister committed:
Added updated composer.lock

manuelmeister committed:
Merge branch 'devel' of github.com:ecamp/ecamp3 into feature/picasso

usu commented:
Einen Eintrag im Picasso mit drag & drop zu verschieben geht bei mir nicht. Folgende Fehlermeldung:
`ServerException: Server-Error 400 (Unrecognized field "startTime")`

DB hab ich schon mal abgeräumt und wieder aufgebaut. Geht das bei dir?

manuelmeister committed:
Fixed wrong api patch

manuelmeister committed:
Removed incorrect success

manuelmeister committed:
Show feedback on patch

manuelmeister commented:
Sollte jetzt klappen. Da war noch eine Überreste von vorher drin.

usu commented:
Cool, Fehler erscheint nicht mehr.

Ansonsten ist etwas schwierig zu verstehen was bereits funktionieren sollte und was nicht funktioniert weils gar noch nicht umgesetzt ist. Ich schreib mal alles auf, und dann können wir am nächsten Meeting sortieren.

### Neue Activity erstellen
- Period Offset ist für den Nutzer nicht verständlich. Nehme ich an, dass hattest du noch vor auf Startzeit zu ändern.
- Wenn ich auf Cancel klicke, bleibt der Platzhalter für die neue Aktivität bestehen. Bewusst?
- Wenn ich submitte wird der Platzhalter entfernt bevor die echte Aktivität erscheint. Kann man das zeitgleich machen?

### Aktivität bearbeiten
- Ich finde keine Möglichkeit, um zur Detailansicht der Aktivität zu gelangen
- Länge im Dialog anpassen geht nicht: Keine Fehlermeldung aber auch kein Change. Wird im Patch-Request gar nicht mitgeschickt.
- Dito mit Offset. Wird im Patch-Request gar nicht mitgeschickt.

### Drag & Drop
Mine Wunschvorstellung wär wie folgt:
- Der User sieht live, was mit seiner Aktivität passiert (Verschieben, verlängern, etc.)
- Sobald der User die Maus loslässt wird die Aktivität im Hintergrund gespeichert
- Alternativ kann man in dieser Zeit die Aktivität sperren und ein Progress-Circle anzeigen. Wenn wir aber einigermassen stabil sind, ist das nicht mal zwingend notwendig und es reicht eine kurze Rückmeldung, wenn das Speichern erfolgreich war (analog wie beim ApiWrapper für Eingabefelder).

manuelmeister commented:
Ja ich bin nicht zufrieden mit meinem Resultat, da klappt noch vieles nicht so wie ich es haben möchte.

Ich hatte gemerkt, dass ich noch weitere Probleme anschneide (zB Dialog Component Embedded Collection Patch, Umrechnung von PeriodOffset auf Starttime).

manuelmeister committed:
Fixed calendar not updating after move and expand

manuelmeister committed:
Don't flash on creation

manuelmeister committed:
Allow BasePicker Picker to have different format than BasePicker TextField

manuelmeister committed:
Added scheduleEntry helpers to handle the case «get the absolute time using the period» everywhere the same

manuelmeister committed:
Allow embeddedCollections in DialogBase

manuelmeister committed:
Show changes in background and delete when cancel

manuelmeister committed:
Move scheduleEntries to embeddedCollections

manuelmeister committed:
Fix scheduleEntries not updating

manuelmeister committed:
Change from periodOffset and length to start and end and process schedule entries with helper to optimize presentation

manuelmeister committed:
Fixed time picker, it now uses unix timestamp in milliseconds as default

manuelmeister committed:
Fixed merge conflicts and added scheduleEntries to activity validator

manuelmeister committed:
Allow scheduleEntries to be updated

manuelmeister committed:
Update picasso on successful update

manuelmeister committed:
Added duration hint and unified entries

manuelmeister committed:
Add button to open existing activity

manuelmeister commented:
@usu Danke fürs Feedback. Ich war mit all deinen Punkten einverstanden und habe es jetzt noch einmal versucht und es hat geklappt. Jetzt funktioniert fast alles (vielleicht noch nicht ganz so schön, zb. mit dem bildschirmfüllenden Dialog, das sollte wohl eher ein `v-menu` sein, aber alles speichert und ist bedienbar.)

Der einzige Fehler im Moment ist, wenn man eine neue Aktivität erstellt, dass dann irgendwie ein andere Entry verloren geht. 😕

usu commented:
> @usu Danke fürs Feedback. Ich war mit all deinen Punkten einverstanden und habe es jetzt noch einmal versucht und es hat geklappt. Jetzt funktioniert fast alles (vielleicht noch nicht ganz so schön, zb. mit dem bildschirmfüllenden Dialog, das sollte wohl eher ein `v-menu` sein, aber alles speichert und ist bedienbar.)

Yeah, sehr cool 👍 
Drag & drop scheint schon mal gut zu klappen. Irgendwas geht bei mir mit den Timepickern nicht richtig, muss mir in ner ruhigen Minute mal anschauen, woran's liegt. Code schau ich dann auch mal noch an.

Ein anderer Punkte (wohl eher für "Meeting-Discuss"): Wenn man einen Entry, verschiebt ändert die Nummerierung ja bei Bedarf nach Erhalt der Server-Antwort. Die Nummern der anderen Einträge bleiben aber natürlich gleich, selbst wenn die Nummerierung nicht mehr stimmt. Könnten wir mal andiskutieren, wie wir damit umgehen...

usu commented on frontend/src/components/dialog/DialogActivityEdit.vue:
```diff
@@ -0,0 +1,79 @@
+    },
+    update () {
+      this.error = null
+      return this.api.patch(this.entityUri, this.entityData).then(this.updatedSuccessful, this.onError)
```
Tut das bei dir? Bei mir friert der Browser hier ein. Ich glaube Axios findet's nicht so lustig, wenn du da ganze JS-Objekte (inkl. Funktionen und so) hineindrückst (=camp & embeddedCollections)

usu commented on backend/module/eCampApi/config/module.config.php:
```diff
@@ -1358,6 +1358,12 @@
                 'filters' => [],
                 'validators' => [],
             ],
+            4 => [
```
Das wird Konflikte haben mit meinem PR #570 
Kann ich aber sonst bereinigen, wenn du magst

usu commented on backend/module/eCampCore/src/Hydrator/ScheduleEntryHydrator.php:
```diff
@@ -44,14 +49,11 @@ public function extract($object) {
             'width' => $scheduleEntry->getWidth(),
 
-            'startTime' => Util::extractDateTime($scheduleEntry->getStartTime()),
-            'endTime' => Util::extractDateTime($scheduleEntry->getEndTime()),
```
Wollen wir die nicht drin lassen? Finde ich eigentlich schon noch user-friendly, das via API gleich zur Verfügung zu stellen.

manuelmeister replied:
## Meeting Notes

Wird zu einem späteren Zeitpunkt geändert. #594

usu commented on backend/module/eCampCore/src/EntityService/ActivityService.php:
```diff
@@ -67,6 +115,36 @@ protected function fetchQueryBuilder($id) {
+        /** @var Activity $activity */
+        $activity = parent::createEntity($data);
+
+        if (isset($data->campId)) {
```
Dito wegen PR #570. Kann ich flicken.

usu commented on backend/module/eCampCore/src/EntityService/ActivityService.php:
```diff
@@ -2,48 +2,96 @@
+        if (isset($data->campId)) {
+            /** @var Camp $camp */
+            $camp = $this->findEntity(Camp::class, $data->campId);
+            $camp->addActivity($activity);
```
Ich glaube nicht, dass wir das erlauben wollen (Camp einer bestehenden Activity ändern).

usu commented on backend/module/eCampLib/src/Hydrator/Util.php:
```diff
@@ -44,24 +46,22 @@ public static function CollectionLink($resolver, $linkResolver) {
         }
 
-        return $date->format('Y-m-d H:i:s');
+        return $date->getTimestamp() * self::MillisecondsInASecond;
```
Weil in Javascript Milliseconds gängig ist?

usu commented on docs/Untitled Diagram.drawio:
```diff
@@ -0,0 +1 @@
+<mxfile host="app.diagrams.net" modified="2020-08-25T18:33:16.964Z" agent="5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36" version="13.6.5" etag="s_OCHXTqF1hKholv-8Q5" type="github"><diagram id="SSDicn0ln8HpbE7rcKx4">UzV2zq1wL0osyPDNT0nNUTV2VTV2LsrPL4GwciucU3NyVI0MMlNUjV1UjYwMgFjVyA2HrCFY1qAgsSg1rwSLBiADYTaQg2Y1AA==</diagram></mxf ...
```
Kann das weg?

usu commented on backend/module/eCampLib/src/Service/UnattachedEntityException.php:
```diff
@@ -0,0 +1,6 @@
+<?php
```
Wird das verwendet?

manuelmeister replied:
Entfernen

usu commented on frontend/src/plugins/scheduleEntries.js:
```diff
@@ -0,0 +1,32 @@
+function defineHelpers (scheduleEntry, timed = false) {
```
Nicht sicher ob der Ordner "plugins" der richtige Ort ist für die helpers. Alles andere sind ja Plugins, welche global bei Vue registriert werden

usu commented on frontend/src/components/form/base/ETimePicker.vue:
```diff
@@ -30,16 +34,60 @@ Allows 15min steps only
   props: {
-    value: { type: String, required: true }
+    icon: { type: String, required: false, default: 'mdi-clock-outline' },
+    value: { type: [Number, String], required: true },
```
Der EDatePicker bräuchte eine ähnliche Ergänzung. Der kann nämlich noch nicht mit Number umgehen und ist damit broken (z.B. Periode editieren).

manuelmeister commented on backend/module/eCampCore/src/Hydrator/PeriodHydrator.php:
```diff
@@ -34,8 +34,8 @@ public function extract($object) {
-            'start' => Util::extractDate($period->getStart()),
-            'end' => Util::extractDate($period->getEnd()),
+            'start' => Util::extractTimestamp($period->getStart()),
+            'end' => Util::extractTimestamp($period->getEnd()),
```
## Meeting Notes
Lieber ISO string da lesbarer.

manuelmeister committed:
Move camp to own prop

manuelmeister committed:
Merge branch 'devel' of github.com:ecamp/ecamp3 into feature/picasso

manuelmeister committed:
Renamed start to periodOffset (merge conflicts)

manuelmeister committed:
Fixed bug, where scheduleEntries are not visible anymore after navigation

manuelmeister committed:
Fix order and optimize parsing

manuelmeister committed:
Allow datepicker to have a different value format

manuelmeister committed:
Removed unused files

manuelmeister committed:
Moved schedule entry helper into components

manuelmeister committed:
Changed period start and end to ISO 8601 RFC3339 format

manuelmeister commented:
This PR is superseded by #595
